rust/ql/lib/codeql/rust/elements/internal/FormatArgumentImpl.qll

```diff
@@ -0,0 +1,86 @@
+/**
+ * This module provides a hand-modifiable wrapper around the generated class `FormatArgument`.
+ *
+ * INTERNAL: Do not use.
+ */
+
+private import codeql.rust.elements.internal.generated.FormatArgument
+private import codeql.rust.elements.internal.generated.Raw
+private import codeql.rust.elements.internal.generated.Synth
+private import codeql.rust.elements.internal.FormatArgumentConstructor
+
+/**
+ * INTERNAL: This module contains the customizable definition of `FormatArgument` and should not
+ * be referenced directly.
+ */
+module Impl {
+  // the following QLdoc is generated: if you need to edit it, do it in the schema file
+  /**
+   * An argument in a format element in a formatting template. For example the `width`, `precision`, and `value` in:
+   * ```rust
+   * println!("Value {value:#width$.precision$}");
+   * ```
+   * or the `0`, `1` and `2` in:
+   * ```rust
+   * println!("Value {0:#1$.2$}", value, width, precision);
+   * ```
+   */
+  class FormatArgument extends Generated::FormatArgument {
+    Raw::FormatArgsExpr parent;
+    int index;
+    int kind;
+    string name;
+    private int offset;
+
+    FormatArgument() {
+      this = Synth::TFormatArgument(parent, index, kind) and
+      formatArgument(parent, index, kind, name, _, offset)
+    }
+
+    override string toString() { result = name }
+
+    override predicate hasLocationInfo(
```

Comment:
Ok, that's possible. Why doesn't the dataflow library work with `hasLocationInfo` instead? That would be a little more generic.

Reply:
It's because we want to reuse [`LocationSig`](https://github.com/github/codeql/blob/226756e1d4a6fcd4822caf48fd139a69e834d0a6/shared/util/codeql/util/Location.qll#L9), instead of having to repeat the `hasLocationInfo` signature in many different places (as well as having to implement `hasLocationInfo` when `getLocation` is already there).

Reply:
I understand, I'm not sure I fully agree though. I think having to define `hasLocationInfo` is less of a burden than having to duplicate the largest table in the database.